### PR TITLE
&yarnabrina [MNT] testing only estimators from modules that have changed compared to `main`

### PR DESCRIPTION
**Pull request by fkiraly**
Experimental PR that ensures only estimators are tested that have changed compared to `main`.

The design relies on a new utility that checks changes to `main` using `subprocess` calling `git`.

How we know it works:
* The CI in this PR runs none of the suite tests for any of the estimators in `sktime`, since this PR does not change any of the relevant files.
* whereas, in this PR https://github.com/sktime/sktime/pull/5023 where two estimator modules have changed, it runs exactly the suite tests for the two estimators.

Also, this most likely gets us a good way towards https://github.com/sktime/sktime/issues/2890, at least when it comes to tests arising from the `TestAll` portion of the test framework.

### CODE REVIEW

**Comment by fkiraly** [comment]
doesn't work like this, errors out with `fatal: bad revision` since it cannot see `main`.

**Comment by fkiraly** [comment]
This seems to be working!

Testing the positive case with https://github.com/sktime/sktime/pull/5023, i.e., the case that two estimator modules change. If that tests exactly those modules with the suite tests, we know it most likely works!

File: sktime/utils/git_diff.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+def is_module_changed(module_str):
+    """Check if a module has changed compared to the main branch.
+
+    Parameters
+    ----------
+    module_str : str
+        module string, e.g., sktime.forecasting.naive
+    """
+    module_file_path = get_path_from_module(module_str)
+    cmd = f"git diff remotes/origin/main -- {module_file_path}"
```
Comment (yarnabrina) [inline sktime/utils/git_diff.py:59]: If you're checking against remote branch, do we need to pull `main`? Can shallow repositories not compare against remote branches?

Comment (yarnabrina) [reply sktime/utils/git_diff.py:59]: May be not, got confused after seeing this: https://stackoverflow.com/questions/23708231/git-shallow-clone-clone-depth-misses-remote-branches

**Review by yarnabrina — COMMENTED** [review]
Thank you so much for this PR. I'll be singificantly happier (at ANY level of significance) if the CI time decreases by running only what's necessary, at least for most of the times (except may be releases?).

I added some comments/doubts/suggestions.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -168,6 +168,8 @@ jobs:
 
     steps:
       - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply .github/workflows/test.yml:172]: I think that checks out *only* main?

File: setup.cfg [hunk 3]
```diff
@@ -19,6 +19,7 @@ addopts =
     --cov-report html
     --showlocals
     --matrixdesign True
+    --only_changed_modules True
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply setup.cfg:22]: does this impact local testing?

File: sktime/utils/git_diff.py (hunk 1 above)
Comment (fkiraly) [reply sktime/utils/git_diff.py:59]: I tried many things, and this is the only one that worked...

I am by no means a `git` expert, so if you have a better idea, feel free to try (preferably in a separate branch unless you are 100% sure it works)

**Comment by fkiraly** [comment]
> at least for most of the times (except may be releases?).

This might actually be a case for a CRON job?

@hazrulakmal suggested this earlier in another context (checking the upstream data repos' health), here's the link he posted:
https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule

File: setup.cfg (hunk 3 above)
[the remark this answers is not in the document]
Comment (yarnabrina) [reply setup.cfg:22]: Yes. I always have to disable pytest extension specific settings from setup.cfg that I don't have installed in my local.

**Comment by yarnabrina** [comment]
> This might actually be a case for a CRON job?

Sure, that'd work. It'll also mean more frequent through testing than what I suggested above as once per new release.

But I do have a question, not specifically related to CRON. If we intend to run tests for all estimators, how shall we instruct CI to disregard `--only_changed_modules True` in `setup.cfg` for only that?

**Comment by fkiraly** [comment]
> If we intend to run tests for all estimators, how shall we instruct CI to disregard `--only_changed_modules True` in `setup.cfg` for only that?

The CRON job uses the flag `--only_changed_modules False`? 

Similar to how we override `--matrixdesign True` to `False`.

File: .github/workflows/test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (fkiraly) [reply .github/workflows/test.yml]: resolved by merging @yarnabrina's proposed solution which fetches only `main`

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -240,10 +252,14 @@ jobs:
     runs-on: ubuntu-20.04
     steps:
       - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply .github/workflows/test.yml]: resolved by merging @yarnabrina's proposed solution which fetches only `main`

File: setup.cfg (hunk 3 above)
[the remark this answers is not in the document]
Comment (fkiraly) [reply setup.cfg:22]: that's a more general issue then - should we open an issue on making the `pytest` setup robust with respect to not all `pytest` extensions being installed?

File: sktime/utils/git_diff.py (hunk 1 above)
Comment (fkiraly) [reply sktime/utils/git_diff.py:59]: I think this is also resolved by adopting @yarnabrina's alternative suggestion by merge to this PR?

Comment (yarnabrina) [reply sktime/utils/git_diff.py:59]: Yes, it's resolved as well.

File: setup.cfg (hunk 3 above)
[the remark this answers is not in the document]
Comment (yarnabrina) [reply setup.cfg:22]: We can, but that'll probably be very difficult as we can not know what everyone has installed. The current `setup.cfg` assumes installation of all test dependencies covered by contribution guide, so then it works fine.

**Review by yarnabrina — APPROVED** [review]

**Pull request merged**